### DISCUSSION TITLE
Error while generating the documentation

### DISCUSSION THREAD

**Discussion by himabindutellakula** [schemaspy/schemaspy#1468] [Q&A]
I am using the SchemaSpy with the SQL Server and I happen to have these errors while generating the documentation.
Any idea why this is happening?

This happened the first time.

Wrote relationship details of 3 tables/views to directory 'Octopus_Documentation' in 2599 seconds.
View the results by opening Octopus_Documentation\index.html
INFO  - Wrote table details in 1 seconds
INFO  - Wrote relationship details of 3 tables/views to directory 'Octopus_Documentation\spg_stg' in 2599 seconds.
INFO  - View the results by opening Octopus_Documentation\spg_stg\index.html
INFO  - Analyzing 'stg'
INFO  - Starting schema analysis
INFO  - Connected to Microsoft SQL Server - 16.00.1000
INFO  - Gathering schema details
Gathering schema details............................................................................................................................................................................................................................................................................................................................................................................................(7sec)
Connecting relationships............................................................................................................................................................................................................................................................................................................................................................................................(2sec)
Writing/graphing summary.INFO  - Gathered schema details in 2 seconds
INFO  - Writing/graphing summary
...Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
Caused by: org.schemaspy.output.html.HtmlException: Failed to generate Orphan diagram
        at org.schemaspy.output.html.mustache.diagrams.OrphanDiagram.writeDiagram(OrphanDiagram.java:80)
        at org.schemaspy.output.html.mustache.diagrams.OrphanDiagram.html(OrphanDiagram.java:56)
        at org.schemaspy.view.HtmlOrphansPage.createPageData(HtmlOrphansPage.java:68)
        at org.schemaspy.view.HtmlOrphansPage.write(HtmlOrphansPage.java:57)
        at org.schemaspy.SchemaAnalyzer.generateHtmlDoc(SchemaAnalyzer.java:424)
        at org.schemaspy.SchemaAnalyzer.analyze(SchemaAnalyzer.java:289)
        at org.schemaspy.SchemaAnalyzer.analyzeMultipleSchemas(SchemaAnalyzer.java:186)
        at org.schemaspy.SchemaAnalyzer.analyze(SchemaAnalyzer.java:127)
        at org.schemaspy.cli.SchemaSpyRunner.runAnalyzer(SchemaSpyRunner.java:109)
        at org.schemaspy.cli.SchemaSpyRunner.run(SchemaSpyRunner.java:98)
        at org.schemaspy.Main.main(Main.java:55)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        ... 5 more
Caused by: org.schemaspy.output.diagram.RenderException: javax.script.ScriptException: Error: renderer for " ALIGN="LEFT"><TR ALIGN="LEFT"><TD ??"LEFT" FIXEDSIZE="TRUE" WIDTH= is unavailable
 in <eval> at line number 94 at column number 4
        at org.schemaspy.output.diagram.vizjs.VizJSDot.render(VizJSDot.java:74)
        at org.schemaspy.output.html.mustache.diagrams.OrphanDiagram.writeDiagram(OrphanDiagram.java:73)
        ... 16 more
		
	
And this happened the second time
		
Wrote relationship details of 14 tables/views to directory 'Octopus_Documentation_20Schemas' in 905 seconds.
View the results by opening Octopus_Documentation_20Schemas\index.html
INFO  - Wrote table details in 15 seconds
INFO  - Wrote relationship details of 14 tables/views to directory 'Octopus_Documentation_20Schemas\census' in 905 seconds.
INFO  - View the results by opening Octopus_Documentation_20Schemas\census\index.html
INFO  - Analyzing 'boe'
INFO  - Starting schema analysis
INFO  - Meta directory "SchemaMeta" should contain a file named "boe.meta.xml"
INFO  - Connected to Microsoft SQL Server - 16.00.1000
INFO  - Gathering schema details
Gathering schema details..........(0sec)
Connecting relationships..........(0sec)
Writing/graphing summary.INFO  - Gathered schema details in 0 seconds
INFO  - Writing/graphing summary
........(13sec)
Writing/diagramming detailsINFO  - Completed summary in 13 seconds
INFO  - Writing/diagramming details
...Exception in thread "main" java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:118)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:51)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:52)
Caused by: org.schemaspy.output.diagram.RenderException: Failed to generate Table diagram
        at org.schemaspy.output.diagram.TableDiagram.generateTableDiagram(TableDiagram.java:24)
        at org.schemaspy.output.html.mustache.diagrams.MustacheTableDiagramFactory.generateRealTableDiagrams(MustacheTableDiagramFactory.java:79)
        at org.schemaspy.output.html.mustache.diagrams.MustacheTableDiagramFactory.generateTableDiagrams(MustacheTableDiagramFactory.java:57)
        at org.schemaspy.SchemaAnalyzer.generateHtmlDoc(SchemaAnalyzer.java:488)
        at org.schemaspy.SchemaAnalyzer.analyze(SchemaAnalyzer.java:289)
        at org.schemaspy.SchemaAnalyzer.analyzeMultipleSchemas(SchemaAnalyzer.java:186)
        at org.schemaspy.SchemaAnalyzer.analyze(SchemaAnalyzer.java:127)
        at org.schemaspy.cli.SchemaSpyRunner.runAnalyzer(SchemaSpyRunner.java:109)
        at org.schemaspy.cli.SchemaSpyRunner.run(SchemaSpyRunner.java:98)
        at org.schemaspy.Main.main(Main.java:55)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        ... 5 more
Caused by: org.schemaspy.output.diagram.RenderException: javax.script.ScriptException: java.lang.StackOverflowError in <eval> at line number 27 at column number 596797
        at org.schemaspy.output.diagram.vizjs.VizJSDot.render(VizJSDot.java:74)
        at org.schemaspy.output.diagram.TableDiagram.generateTableDiagram(TableDiagram.java:21)
        ... 15 more

**Comment by npetzall** [comment]
Hi, I'm guessing your trying to document a big database. The first error I'm a bit unsure about, but it might be related to the second one.

The second one is a known issue since it's a big database you might need to increase the stack size in java, #462 

The first one could be encoding, but I need to see if I can spot the places in our code where we create the surrounding markup so that I can reduce the list of suspects.

But as a workaround for the second exception increase stack size as mentioned in the issue, or switch to graphviz instead of vizjs, unfortunately that comes with its own caveats and edge cases. If docker images is possible that is your safest bet.

**Comment by himabindutellakula** [comment]
Increasing the stack size fixed the errors. But I can see all these errors 

INFO  - Analyzing 'wb'
INFO  - Starting schema analysis
INFO  - Connected to Microsoft SQL Server - 16.00.1000
INFO  - Gathering schema details
Gathering schema details.....(0sec)
Connecting relationships.....(0sec)
Writing/graphing summary.INFO  - Gathered schema details in 0 seconds
INFO  - Writing/graphing summary
INFO  - Graphviz renderer set to ':cairo'
.ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: in label of node wb_master_spread
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: in label of node wb_call_dates
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.compact.dot -orelationships.real.compact.png -Tcmapx: in label of node bloom.bloom_security
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content
ERROR - Graphviz-11.0.0-win64\bin\dot -Tpng:cairo relationships.real.large.dot -orelationships.real.large.png -Tcmapx: Warning: cell size too small for content

**Reply by npetzall** [reply]
We have discussed changing this since people are missing that it's mentioned as warning in the message. We are executing graphviz and I think this is captured from the stderr output hence the error logging, but the messages are warning.

So it's warnings and not errors.

You`ve switched to using graphviz which makes the stack size irrelevant was only applicable when using vizjs (JavaScript implementation run inside of the java)

**Comment by himabindutellakula** [comment]
And I can see duplicate referencing in case of views.
Let us say I have a view named agency_data_vw and when I am on that view page, in the section **Possibly Referenced Tables/Views** it is showing the reference to the same view again.

And one more issue is that, let us say this view agency_data_vw gets the data from the table agency.base_info, 
I cannot see the listing of this source table  agency.base_info  user the section of **Possibly Referenced Tables/Views**  of agency_data_vw page.

Is this a known error? or am I missing something here?

**Reply by npetzall** [reply]
Yes and no, we have no reliable way of getting view information so it`s only based on checking the definition of a view, since the syntax may differ between vendors. It's a naive implementation just checking if any known tables or views occur in the definition.

If not to sensitive, could you provide the definition än we can look into improving the identification.

We could probably remove the view itself should be as the name in the definition and I'm not sure if self referencing views are a thing but assume that it's possible, however you wouldn't need a link back to the page your at.

But need to ponder that a bit.

**Reply by himabindutellakula** [reply]
Here is a sample of the view definition 

CREATE VIEW [data].[agency_data_vw ]
AS
SELECT
    bucket_name, volume, beta, r
FROM agency.base_info WITH (NOLOCK);

I am expecting a link to the base_info table under the section of Possibly Referenced Tables/Views of agency_data_vw page

**Reply by npetzall** [reply]
That looks like it should identify it. I'll have look might be considered a bug.

**Reply by himabindutellakula** [reply]
Right! And a link to the same view page is kind of misleading.

**Comment by himabindutellakula** [comment]
Is it mentioned any where in the documentation about what are minimum database permissions required for the user running schemaspy?

And do I have to create a SchemaMeta file for every schema manually, if I want to store comments there?

**Reply by npetzall** [reply]
There is no such generic documentation, mainly because permissions differ between dbms in general. But there are also custom queries for some of the dbms. But a general rule would be read access to the schema but also information_schema.

Generally if you have the ability to modify the schema I would add the comments there instead of SchemaMeta.

I'm not 100% certain but I think there is a big overlap between the SchemaMeta format and the xml output that is produced.

It would be a good feature improvement if it was the same or a small xslt could be written to convert it to a basic xml with only comments.

**Comment by himabindutellakula** [comment]
I have a requirement to store the comments/description at some other source other than schema or extended properties(in case of SQL Server). And this is because they will be lost if someone drops and creates the table again. In such a case SchemaMeta XML is a good place to store them.
But the problem here is we have to manually prepare and add that file for every schema before generating the documentation. And considering a database with around 100 schemas and 2000 tables and so many columns, creating that XML files is a very tiring process for sure.

When I saw this 
![image](https://github.com/schemaspy/schemaspy/assets/153971509/72449f61-4aa1-4cdc-95c4-9abbdc41f315)
I assumed that SchemaSpy would at least generate a skeleton SchemaMeta file for all the metadata, so that the user can just add comments to it. But I was wrong.
